src/libPMacc/include/eventSystem/transactions/Transaction.tpp

```diff
@@ -60,15 +47,43 @@ inline EventTask Transaction::getTransactionEvent( )
     return baseEvent;
 }
 
-void Transaction::operation( ITask::TaskType )
+void Transaction::operation( ITask::TaskType operation )
 {
-    if ( isAtomic == false )
-        baseEvent.waitForFinished( );
+    if ( operation == ITask::TASK_CUDA )
+    {
+        Manager &manager = Environment<>::get( ).Manager( );
+
+        ITask* baseTask = manager.getITaskIfNotFinished( this->baseEvent.getTaskId( ) );
+        if ( baseTask != NULL )
+        {
+            if ( baseTask->getTaskType( ) == ITask::TASK_CUDA )
+            {
+                /* no blocking is needed */
+                return;
+            }
+        }
+    }
+    baseEvent.waitForFinished( );
 }
 
 EventStream* Transaction::getEventStream( ITask::TaskType )
 {
-    return this->eventStream;
+    Manager &manager = Environment<>::get( ).Manager( );
+    ITask* baseTask = manager.getITaskIfNotFinished( this->baseEvent.getTaskId( ) );
+
+    if ( baseTask != NULL )
+    {
+        if ( baseTask->getTaskType( ) == ITask::TASK_CUDA )
+        {
+            /* `StreamTask` from previous task must be reused to guarantee
+             * that the dependency chain not brake
```

Comment:
Yes `TaskLogicalAnd` has two task if it is created. Based on the two tasks with a streams one stream is selected as *master* and the other one is synchronized to *master*. The event of `TaskLogicalAnd` is only called if one of the two events finished then `TaskLogicalAnd` inherit the stream and the event from this one which still exists.